org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/TryWithResourcesJavacFilter.java

```diff
@@ -0,0 +1,245 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Filters code that javac generates for try-with-resources statement.
+ */
+public final class TryWithResourcesJavacFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterOutput output) {
+		if (methodNode.tryCatchBlocks.isEmpty()) {
+			return;
+		}
+		final Matcher matcher = new Matcher(output);
+		for (TryCatchBlockNode t : methodNode.tryCatchBlocks) {
+			if ("java/lang/Throwable".equals(t.type)) {
+				for (Matcher.JavacPattern p : Matcher.JavacPattern.values()) {
+					matcher.start(t.handler);
+					if (matcher.matchJavac(p)) {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	static class Matcher {
+		private final IFilterOutput output;
+
+		private final Map<String, VarInsnNode> vars = new HashMap<String, VarInsnNode>();
+		private String expectedOwner;
+
+		private AbstractInsnNode start;
+		private AbstractInsnNode cursor;
+
+		Matcher(final IFilterOutput output) {
+			this.output = output;
+		}
+
+		private enum JavacPattern {
+			OPTIMAL, FULL, OMITTED_NULL_CHECK, METHOD,
+		}
+
+		private void start(final AbstractInsnNode start) {
+			this.start = start;
+			cursor = start.getPrevious();
+			vars.clear();
+			expectedOwner = null;
+		}
+
+		private boolean matchJavac(final JavacPattern p) {
+			// "catch (Throwable t)"
+			if (!nextIsVar(Opcodes.ASTORE, "t1")) {
+				return false;
+			}
+			// "primaryExc = t"
+			if (!nextIsVar(Opcodes.ALOAD, "t1")) {
+				return false;
+			}
+			if (!nextIsVar(Opcodes.ASTORE, "primaryExc")) {
+				return false;
+			}
+			// "throw t"
+			if (!nextIsVar(Opcodes.ALOAD, "t1")) {
+				return false;
+			}
+			if (!nextIs(Opcodes.ATHROW)) {
+				return false;
+			}
+
+			// "catch (any t)"
+			if (!nextIsVar(Opcodes.ASTORE, "t2")) {
+				return false;
+			}
+			if (!nextIsJavacClose(p, "e")) {
+				return false;
+			}
+			// "throw t"
+			if (!nextIsVar(Opcodes.ALOAD, "t2")) {
+				return false;
+			}
+			if (!nextIs(Opcodes.ATHROW)) {
+				return false;
+			}
+			final AbstractInsnNode end = cursor;
+
+			AbstractInsnNode startOnNonExceptionalPath = start.getPrevious();
+			cursor = startOnNonExceptionalPath;
+			while (!nextIsJavacClose(p, "n")) {
+				startOnNonExceptionalPath = startOnNonExceptionalPath
+						.getPrevious();
+				cursor = startOnNonExceptionalPath;
+				if (cursor == null) {
```

Comment:
@marchof maybe can get reduce amount of branches by changing `nextIs` and `nextIsVar` methods: if cursor is `null`, then do nothing, instead of returning `false` set cursor to `null`, the loop to find closing of resource on non-exceptional path already contains null-check. However question of performance of such approach is quite interesting: will JVM be able to optimize all this by inlining in order to retain early exits in case of missmatches? if not, then what will be cost of unnecessary monomorphic method calls? null-check supposed to be cheap as JVM anyway need to do it during field access. WDYT?